main.js

```diff
@@ -1,129 +1,88 @@
+const { defineLazyGetter } = devtoolsRequire("devtools/toolkit/DevToolsUtils");
+
+const events = require("sdk/event/core");
+const { when: unload } = require("sdk/system/unload");
+const { Devices } = Cu.import("resource://gre/modules/devtools/Devices.jsm");
+const { gDevToolsBrowser } =
+  Cu.import("resource:///modules/devtools/gDevTools.jsm");
+defineLazyGetter(this, "adb", () => {
+  return require("./adb");
+});
+defineLazyGetter(this, "fastboot", () => {
+  return require("./fastboot");
+});
+defineLazyGetter(this, "Device", () => {
+  return require("./device");
+});
+
+// Set this right away on startup
+Devices.helperAddonInstalled = true;
+unload(() => Devices.helperAddonInstalled = false);
 
-let promise;
-try {
-  promise = Cu.import("resource://gre/modules/commonjs/promise/core.js").Promise;
-} catch (e) {
-  promise = Cu.import("resource://gre/modules/commonjs/sdk/core/promise.js").Promise;
+// In Firefox 39+ startup can be deferred until WebIDE is open
+if (gDevToolsBrowser.isWebIDEInitialized) {
+  gDevToolsBrowser.isWebIDEInitialized.promise.then(start);
+} else {
+  start();
 }
 
-// As of Firefox 36, WebIDE exposes an API to register new runtimes.
-try {
+// Main startup sequence to listen for new devices, inform WebIDE, etc.
+function start() {
+  // As of Firefox 36, WebIDE exposes an API to register new runtimes.
   let Runtimes = devtoolsRequire("devtools/webide/runtimes");
   if (Runtimes && Runtimes.RuntimeScanners) {
     let scanner = require("./scanner");
     scanner.register();
   }
-} catch (e) {}
-
-Devices.helperAddonInstalled = true;
-exports.shutdown = function() {
-  Devices.helperAddonInstalled = false;
-  adb.stop(true);
-};
-
-adb.start().then(function () {
-  adb.trackDevices();
-});
 
-/**
- * A Device instance is created and registered with the Devices module whenever
- * ADB notices a new device is connected.
- *
- * Any changes here should be examined carefully for backwards compatibility
- * issues.  Other add-ons, like the Tools Adapter, make use of these low-level
- * device objects too.
- */
-function Device(id) {
-  this.id = id;
-}
+  adb.start().then(function() {
+    adb.trackDevices();
+  });
+  unload(() => adb.stop(true));
 
-Device.prototype = {
   /**
-   * DEPRECATED: This is specific to how we connect to Firefox OS.  Use cases
-   * that interact with other kinds of devices should likely use the more
-   * general |forwardPort| method directly.
+   * It may seem at first that registering devices with Device.jsm is no longer
+   * needed, since they are now consumed in this same add-on in scanner.js (at
+   * least for Firefox 36+ with new WebIDE API).  However, there are still use
+   * cases for the Devices.jsm registry, as other add-ons can make use of these
+   * low-level ADB devices if they know about other new runtimes to support on
+   * the device.  This the approach used in the Fever Dream / Valence add-on to
+   * find Chrome on Android, for example.
    */
```

Comment:
It may have more sense to put this comment just before `onConnected` and its `Devices.register` usage.